xcube/core/store/preload.py

```diff
@@ -0,0 +1,409 @@
+    def cancelled(self) -> bool:
+        """Return true if and only if the internal flag is true."""
+        return self._cancel_event.is_set()
+
+    def cancel(self):
+        self._cancel_event.set()
+        for future in self._futures.values():
+            future.cancel()
+        self._executor.shutdown(wait=False)
+
+    def notify(self, event: PreloadState):
+        state = self._states[event.data_id]
+        if (
+            event.status is not None
+            and event.status != state.status
+            and state.status
+            in (PreloadStatus.stopped, PreloadStatus.cancelled, PreloadStatus.failed)
+        ):
+            # Status cannot be changed
+            return
+        with self._lock:
+            state.update(event)
+            self._display.update()
+
+    def preload_data(self, data_id: str):
+        """Preload the data resource given by *data_id*.
+
+        Concurrently executes the *preload_data* passed to the constructor,
+        if any. Otherwise, it does nothing.
+
+        Can be overridden by clients to implement the actual preload operation.
+
+        Args:
+           data_id: The data identifier of the data resource to be preloaded.
+        """
+        if self._preload_data is not None:
+            self._preload_data(self, data_id)
+
+    def _run_preload_data(self, data_id: str) -> str:
+        self.notify(PreloadState(data_id, status=PreloadStatus.started))
+        self.preload_data(data_id)
+        return data_id
+
+    def _handle_preload_data_done(self, f: Future[str]):
+        # Find the data_id that belongs to given feature f
+        data_id: str | None = None
+        for data_id, future in self._futures.items():
+            if f is future:
+                break
+        assert data_id is not None
+        try:
+            _value = f.result()
+            # No exceptions, notify everything seems ok
+            self.notify(PreloadState(data_id, status=PreloadStatus.stopped))
+        except CancelledError as e:
+            # Raised if future has been cancelled
+            # while executing `_run_preload_data`
+            self.notify(
+                PreloadState(data_id, status=PreloadStatus.cancelled, exception=e)
+            )
+        except Exception as e:
+            # Raised if any exception occurred
+            # while executing `_run_preload_data`
+            self.notify(PreloadState(data_id, status=PreloadStatus.failed, exception=e))
+
+    def show(self) -> Any:
+        return self._display.show()
+
+    def _repr_html_(self):
+        return self._display.to_html()
+
+    def __str__(self):
+        return self._display.to_text()
+
+    def __repr__(self):
+        return self._display.to_text()
+
+    def __enter__(self) -> "PreloadHandle":
+        """Enter the context.
+
+        Does nothing but returning this handle.
+        Only useful when in blocking mode.
+
+        Returns:
+            This object.
+        """
+        return self
```

Comment:
This is the same as the base class, can be removed.

Reply:
Yepp.